permission-element.bs

```diff
@@ -19,7 +18,408 @@ Abstract: A `<permission>` HTML element to request browser permissions in-page.
+1. [=Add a blocker=] with |reason| and false.
+
+</div>
+
+<div algorithm>
+To <dfn for="HTMLPermissionElement">add a permament blocker</dfn> with a
+{{PermissionElementInvalidReason}} |reason|:
+
+1. [=Assert=]: |reason| is listed as "permanent" in the [=blocker reason table=].
+1. [=Add a blocker=] with |reason| and false.
+
+</div>
+
+<div algorithm>
+To <dfn for="HTMLPermissionElement">remove blockers</dfn> with
+{{PermissionElementInvalidReason}} |reason| from an |element|:
+
+1. [=Assert=]: |reason| is listed as "temporary" in the [=blocker reason table=].
+1. [=list/iterate|For each=] |entry| in |element|'s {{[[BlockerList]]}}:
+    1. If |entry|'s reason [=string/is|equals=] |reason|, then [=list/remove=]
+        |entry| from |element|'s {{[[BlockerList]]}}.
+
+</div>
+
+<div algorithm>
+To determine a {{HTMLPermissionElement}} |element|'s
+<dfn for="HTMLPermissionElement">blocker</dfn>:
+
+1. [=list/iterate|For each=] |entry| in |element|'s {{[[BlockerList]]}}:
+    1. If |entry| is [=HTMLPermissionElement/valid=], then return |entry|.
+1. Return nothing.
+
+</div>
+
+<div algorithm>
+An {{HTMLPermissionElement}}'s [=blocker=] list's |entry| is
+<dfn for="HTMLPermissionElement">valid</dfn> if:
+
+1. |entry| has no timestamp,
+1. or |entry| has a timestamp, and the timestamp is greater or equal to the
+    [=current high resolution time=].
+
+</div>
+
+NOTE: The spec does not define a way for expiring blockers that have expired
+    to be removed, so {{[[BlockerList]]}} may grow indefinitely.
+    Since the presence of expired blockers is not observable, user agents
+    should feel free to clean up the blocker list.
+
+## <{permission}> element algorithms ## {#permission-element-algorithms}
+
+<div algorithm="HTMLPermissionElement/constructor">
+The {{HTMLPermissionElement}} constructor steps are:
+
+1. Initialize the internal {{[[Types]]}} slot to null.
+1. Initialize the internal {{[[BlockerList]]}} to &laquo;[]&raquo;.
+
+</div>
+
+<div algorithm="HTMLPermissionElement/insertion steps">
+The {{HTMLPermissionElement}} [=insertion steps=] are:
+
+1. If {{[[Types]]}} is null, set {{[[Types]]}} to &laquo;[]&raquo;.
+1. Initialize the internal {{[[BlockerList]]}} to &laquo;[]&raquo;
+1. Initialize the internal {{[[IntersectionObserver]]}} with the result of
+    constructing a new {{IntersectionObserver}}, with
+    [=HTMLPermissionElement/IntersectionObserver callback=].
+1. Call {{[[IntersectionObserver]]}}.observe([=this=]).
+1. [=Add an expiring blocker=] with reason `"recently_created"`.
+
+</div>
+
+<div algorithm="HTMLPermissionElement/isValid">
+HTMLPermissionElement |element|'s <dfn>isValid</dfn> getter steps are:
+
+1. Return whether |element|'s [=HTMLPermissionElement/blocker=] is Nothing.
+
+</div>
+
+<div algorithm="HTMLPermissionElement/invalidReason">
+HTMLPermissionElement |element|'s <dfn>invalidReason</dfn> getter steps are:
+
+1. If |element| {{HTMLPermissionElement/isValid}}, then return
```

Comment:
If isValid is true this should return "" and only return blocker's reason string if isValid is false right?

Reply:
Ha, yes. Fixed.